### PR TITLE
don't use memcmp() for struct equality, do memberwise comparisons per the spec

### PR DESCRIPTION
**Pull request by WalterBright**
I.e. if the structs fit in 2 or fewer registers use a member-wise compare.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by WalterBright** [comment]
Hmm. Needs more work.

**Comment by WalterBright** [comment]
Who knows why these tests are failing. Can't tell anything from the logs. For one thing, the logs should STOP when there's a failure, not continue going on for thousands of lines. The thing that failed should be the last line. The tests in test/runnable are all simple to diagnose with simple logs.

**Comment by Geod24** [comment]
> the logs should STOP when there's a failure, not continue going on for thousands of lines

True, but not trivial to do, especially if we want to run tests concurrently.

> Who knows why these tests are failing.

There seems to be multiple reason. One I was able to dig out pretty easily on Windows is:

> std\variant.d(270): Warning: statement is not reachable

[Handy link to L270](https://github.com/dlang/phobos/blob/e10d6a7e973ed85ccceee21088fa716433429583/std/variant.d#L270)

**Comment by WalterBright** [comment]
Can't repro the variant.d issue, either on Linux or Windows. Sigh. Looking at the code in variant.d, I don't see any reason for the warning, either. Can anyone else repro a problem? It's only a one line change to the compiler.

**Comment by WalterBright** [comment]
@schveiguy @wilzbach as recent committers to std.variant, can you perhaps help out with what is going on here?

The one line change to the compiler replaces a bit compare of struct equality with a member-wise compare for some structs.

**Comment by WalterBright** [comment]
Spec reference: https://github.com/dlang/dlang.org/pull/2572

**Comment by WalterBright** [comment]
Revised the PR so it is in conformance with the spec.

**Comment by WalterBright** [comment]
I think this will need a `-preview=fieldcmp` switch, because it is apparently breaking existing code. But that is blocked by https://github.com/dlang/dmd/pull/9206

**Comment by WalterBright** [comment]
Rather than wait for #9206, I just went ahead and added a `-preview=fieldwise` switch directly and unblocked this.

File: changelog/no-cmpsb.dd
Comment (jacob-carlborg) [inline changelog/no-cmpsb.dd]: This will add a literal `#` at the beginning of the line, see: http://dtest.dlang.io/artifact/website-6d079ba9822f488660df930751bffd99a564cbfd-9b21681f7b99def3027cda434bb08739/web/changelog/pending.html#no-cmpsb.

**Comment by aG0aep6G** [comment]
Related issue: [16216](https://issues.dlang.org/show_bug.cgi?id=16216).

File: changelog/no-cmpsb.dd
Comment (wilzbach) [inline changelog/no-cmpsb.dd]: FWIW this link won't be clickable because Ddoc doesn't support this. You need to use LINK2 or similar macros.

Comment (WalterBright) [reply changelog/no-cmpsb.dd]: Ddoc should be auto-recognizing URLs.

File: changelog/no-cmpsb.dd
Comment (WalterBright) [reply changelog/no-cmpsb.dd]: I thought Ddoc did markdown now.

Comment (wilzbach) [reply changelog/no-cmpsb.dd]: It's still not activated by default ;-)

**Comment by WalterBright** [comment]
@aG0aep6G Thank you, I wasn't aware of that bug report. This should fix it (but not for `is` comparisons).

**Comment by WalterBright** [comment]
@wilzbach @thewilsonator ready to pull now.

File: changelog/no-cmpsb.dd
Comment (wilzbach) [inline changelog/no-cmpsb.dd]: This links stands a bit out in the flow of the reader:

image


Maybe directly link it as e,g,  "[Equality Expressions](...) can break ..."?

**Review by wilzbach — APPROVED** [review]
LGTM.
Though I want to point out that we shouldn't abuse the new `-preview`  too much as we still aren't planning to do a D3 (or D 2020 (?)) in the foreseeable future (?) and thus breaking changes must still go through a deprecation phase in which the users can still compile their code, but are warned about the change.
Just by adding a `-preview` switch a user won't be warned that this code is affected by this breaking change.

**Comment by wilzbach** [comment]
> True, but not trivial to do, especially if we want to run tests concurrently.

Isn't this more a "feature" that someone actively worked into the testsuite?
I think the idea was that it takes quite a long time to get feedback from all machines, so you at least want to get a full list of all failed tests.
Though I absolutely agree that all failed tests should be printed as a summary at the end.
If we agree to switch over to `test/run.d` on the auto-tester this could be added easily...

> @schveiguy @wilzbach as recent committers to std.variant, can you perhaps help out with what is going on here?
> The one line change to the compiler replaces a bit compare of struct equality with a member-wise compare for some structs.

Git blame points to five years ago:
 https://github.com/dlang/phobos/blame/e10d6a7e973ed85ccceee21088fa716433429583/std/variant.d#L270

- https://github.com/dlang/phobos/pull/1963
- https://github.com/dlang/phobos/pull/1963/commits/83b0cf52a5712fa1d1c2aaf984bbf31390f6b35f

**Comment by WalterBright** [comment]
The problem is I don't know how to detect that the user code would break with this. The user code that relies on the old behavior is wrong in any case. Also, the `-revert` switch will save those who suspect the new behavior is causing them problems.

**Comment by jacob-carlborg** [comment]
> Just by adding a `-preview` switch a user won't be warned that this code is affected by this breaking change.

Before this will be turned on by default it needs to go through a deprecation period.

**Comment by jacob-carlborg** [comment]
> If we agree to switch over to `test/run.d` on the auto-tester this could be added easily...

My unit test runner already does this 😃 [1].

[1] https://github.com/dlang/dmd/pull/9333

**Pull request merged**
